c_src/dpiStmt_nif.c

```diff
@@ -0,0 +1,507 @@
+			return RAISE_EXCEPTION("Bad data argument");
+		dat->value.asBytes.ptr = bindbuf;
+		dat->value.asBytes.length = strlen(bindbuf);
+		break;
+	case DPI_NATIVE_TYPE_BOOLEAN:
+		if (!enif_get_atom(
+				env, dataTerm, bindbuf, 256, ERL_NIF_LATIN1)) // recycling bindbuf
+			return RAISE_EXCEPTION("Bad data argument");
+		int boolean;
+		if (!strcmp(bindbuf, "false"))
+			boolean = 0;
+		else if (!strcmp(bindbuf, "true"))
+			boolean = 1;
+		else
+			return RAISE_EXCEPTION("Bad data argument");
+		break;
+	case DPI_NATIVE_TYPE_TIMESTAMP: // TODO test
+		ERL_NIF_TERM mapValue;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "fsecond"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		ErlNifUInt64 fsecond;
+		enif_get_uint64(env, mapValue, &fsecond);
+		//dat.value.asTimestamp.fsecond = fsecond;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "second"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		unsigned int second;
+		enif_get_uint(env, mapValue, &second);
+		//dat.value.asTimestamp.second = second;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "minute"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		unsigned int minute;
+		enif_get_uint(env, mapValue, &minute);
+		//dat.value.asTimestamp.minute = minute;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "hour"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		unsigned int hour;
+		enif_get_uint(env, mapValue, &hour);
+		//dat.value.asTimestamp.hour = hour;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "day"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		unsigned int day;
+		enif_get_uint(env, mapValue, &day);
+		//dat.value.asTimestamp.day = day;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "month"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		unsigned int month;
+		enif_get_uint(env, mapValue, &month);
+		//dat.value.asTimestamp.month = month;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "year"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		int year;
+		enif_get_int(env, mapValue, &year);
+		//dat.value.asTimestamp.year = year;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "tzMinuteOffset"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		int tzMinuteOffset;
+		enif_get_int(env, mapValue, &tzMinuteOffset);
+		//dat.value.asTimestamp.tzMinuteOffset = tzMinuteOffset;
+
+		if (!enif_get_map_value(
+				env, dataTerm, enif_make_atom(env, "tzHourOffset"), &mapValue))
+			return RAISE_EXCEPTION("Bad data argument");
+		int tzHourOffset;
+		enif_get_int(env, mapValue, &tzHourOffset);
+		//dat.value.asTimestamp.tzHourOffset = tzHourOffset;
+		dpiData_setTimestamp(dat, year, month, day, hour, minute, second,
+							 fsecond, tzHourOffset, tzMinuteOffset);
```

Comment:
**IMPORTANT** _there shouldn't be any `dpi*` function calls other than `dpiStmt` in `dpiStmt_nif.c`_